storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java

```diff
@@ -0,0 +1,2091 @@
+
+        verifyRecordValidationStats(validatedResults.recordValidationStats, 3, records, true);
+    }
+
+    private void checkRecompression(byte magic) {
+        long now = System.currentTimeMillis();
+        // Set the timestamp of seq(1) (i.e. offset 1) as the max timestamp
+        List<Long> timestampSeq = Arrays.asList(now - 1, now + 1, now);
+
+        long producerId;
+        short producerEpoch;
+        int baseSequence;
+        boolean isTransactional;
+        int partitionLeaderEpoch;
+
+        if (magic >= RecordBatch.MAGIC_VALUE_V2) {
+            producerId = 1324L;
+            producerEpoch = 10;
+            baseSequence = 984;
+            isTransactional = true;
+            partitionLeaderEpoch = 40;
+        } else {
+            producerId = RecordBatch.NO_PRODUCER_ID;
+            producerEpoch = RecordBatch.NO_PRODUCER_EPOCH;
+            baseSequence = RecordBatch.NO_SEQUENCE;
+            isTransactional = false;
+            partitionLeaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH;
+        }
+
+        MemoryRecords records = MemoryRecords.withRecords(
+                magic,
+                0L,
+                Compression.NONE,
+                TimestampType.CREATE_TIME,
+                producerId,
+                producerEpoch,
+                baseSequence,
+                partitionLeaderEpoch,
+                isTransactional,
+                new SimpleRecord(timestampSeq.get(0), "hello".getBytes()),
+                new SimpleRecord(timestampSeq.get(1), "there".getBytes()),
+                new SimpleRecord(timestampSeq.get(2), "beautiful".getBytes())
+        );
+
+        // V2 has single batch, and other versions have many single-record batches
+        assertEquals(magic >= RecordBatch.MAGIC_VALUE_V2 ? 1 : 3, iteratorSize(records.batches().iterator()));
+
+        LogValidator.ValidationResult validatingResults = new LogValidator(
+                records,
+                topicPartition,
+                time,
+                CompressionType.NONE,
+                Compression.gzip().build(),
+                false,
+                magic,
+                TimestampType.CREATE_TIME,
+                1000L,
+                1000L,
+                partitionLeaderEpoch,
+                AppendOrigin.CLIENT,
+                MetadataVersion.latestTesting()
+        ).validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(0L),
+                metricsRecorder,
+                RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        );
+
+        MemoryRecords validatedRecords = validatingResults.validatedRecords;
+
+        int i = 0;
+        for (RecordBatch batch : validatedRecords.batches()) {
+            assertTrue(batch.isValid());
+            assertEquals(TimestampType.CREATE_TIME, batch.timestampType());
+            maybeCheckBaseTimestamp(timestampSeq.get(0), batch);
+            assertEquals(batch.maxTimestamp(), iterToStream(batch.iterator())
```

Comment:
`iterToStream` is used to find the max timestamp, so we can enhance it to return max timestamp directly